FreeRTOS/FreeRTOS-Kernel
# Tickless idle fixes/improvement

Fixes/Improvement to Tickless Idle

Description
-----------
The default implementation of tickless idle on Cortex M has two minor edge-case bugs as identified in this forum thread: https://forums.freertos.org/t/ultasknotifytake-timeout-accuracy/9629.

Both bugs induce a "jump ahead" in xTickCount compared to time as measured by the SysTick.  The first bug (when stopping SysTick on zero) jumps ahead one full tick.  The second bug (when rescheduling an imminent tick after an early wakeup) jumps ahead *nearly* one full tick.

Note that the proposed fix for the first bug also includes an enhancement for tickless idle not to stop SysTick at all in the case `eTaskConfirmSleepModeStatus()` returns eAbortSleep.

Configurations that use the SysTick's alternate clocking have additional, related issues.  In these configurations, xTickCount might jump ahead by the entire expected idle time if the alternate clock is slow.  See https://forums.freertos.org/t/xeventgroupwaitbits-premature-timeout-cause/10432.  This pull request fixes those issues too.

Test Steps
-----------
These bugs are difficult to reproduce due to critical timing.  Could use some help here.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
